File: backend/internal/ports/driven/repository/permission.go
package portsdrivenrepository

import (
	"context"
	"encoding/json"

	domainmodels "github.com/uwb-ips-dteti-ugm/ips-app/backend/internal/domain/models"
)

type Permission interface {
	CreatePermission(ctx context.Context, name string, description string, createdBy *int64) (id int64, err error)
	ReadPermissionById(ctx context.Context, id int64) (permission *domainmodels.Permission, err error)
	ReadPermissionsByPagination(ctx context.Context, page int, limit int, cursorId *int64, search *string) (permissions []domainmodels.Permission, total int, err error)
	UpdatePermissionById(ctx context.Context, id int64, name *string, description *string, updatedBy *int64) (err error)
	UpdatePermissionPreferencesById(ctx context.Context, id int64, preferences json.RawMessage, updatedBy *int64) (err error)
	DeletePermissionById(ctx context.Context, id int64) (err error)
}

File: backend/src/ips_app/domain/contracts/repository/permission.py
from abc import ABC, abstractmethod
from typing import Any, Dict, List, Optional, Tuple

from ips_app.domain.models.permission import Permission


class PermissionRepository(ABC):
    @abstractmethod
    async def create_permission(
        self,
        name: str,
        description: str = "",
        created_by: Optional[Any] = None,
        session: Optional[Any] = None,
    ) -> Permission: ...

    @abstractmethod
    async def read_permission_by_id(
        self,
        id: Any,
        session: Optional[Any] = None,
    ) -> Permission: ...

    @abstractmethod
    async def read_permission_by_name(
        self,
        name: str,
        session: Optional[Any] = None,
    ) -> Permission: ...

    @abstractmethod
    async def read_permissions_by_pagination(
        self,
        page: int,
        limit: int,
        search: Optional[str] = None,
        session: Optional[Any] = None,
    ) -> Tuple[List[Permission], int]: ...

    @abstractmethod
    async def update_permission_by_id(
        self,
        id: Any,
        name: Optional[str] = None,
        description: Optional[str] = None,
        updated_by: Optional[Any] = None,
        session: Optional[Any] = None,
    ) -> Permission: ...

    @abstractmethod
    async def update_permission_preferences_by_id(
        self,
        id: Any,
        preferences: Dict[str, Any],
        updated_by: Optional[Any] = None,
        session: Optional[Any] = None,
    ) -> Permission: ...

    @abstractmethod
    async def delete_permission_by_id(
        self,
        id: Any,
        session: Optional[Any] = None,
    ) -> None: ...
